kymotracker: add `split` action

JoepVanlier opened the pull request:
**Why this PR?**
Nothing too exciting, but requested more than once, this PR adds a split action to the kymotracker.

![split_tracks](https://user-images.githubusercontent.com/19836026/232106238-b3335c43-9da9-4d52-b653-2d7620629f2c.gif)

Note that I piggybacked a few improvements for the localization data on this PR, which allows us to not lose the Gaussian refinement metadata if it exists.

rpauszek commented on changelog.md:
```diff
@@ -5,6 +5,7 @@
 #### New features
 
 * Added [`DwelltimeModel.profile_likelihood()`](https://lumicks-pylake.readthedocs.io/en/latest/_api/lumicks.pylake.DwelltimeModel.html#lumicks.pylake.DwelltimeModel.profile_likelihood) for calculating profile likelihoods for dwell time analysis. These can be used to determine confidence intervals and assess whether the model has too many parameters (i.e., if a model with fewer components would b ...
+* Added functionality to split lines to the Kymotracking widget.
```
```suggestion
* Added functionality to split tracks in the Kymotracking widget.
```

rpauszek commented on changelog.md:
```diff
@@ -18,6 +19,7 @@
 * Added lower and upper bounds when fitting dwell time models. This prevents numerical issues that occur when approaching these extremes. Prior to this change, over-parameterized models could encounter numerical issues due to divisions by (near) zero. After this change, amplitudes are constrained to stay between `1e-9` and `1.0 - 1e-9`, while lifetimes are constrained to remain between `1e-8` and ...
 * Provided better initial guess for dwell time models with three or more components. Prior to this change, the dwelltime models involving three components or more would get an initial guess with an average lifetime beyond the mean of the observations.
 * Silenced false positive warnings about the optimization violating the bound constraints. Prior to this change, dwell time analysis would produce warnings about the optimizer exceeding bounds. These excursions beyond the bound are very small and have no effect on the model.
+* Parameters corresponding to Gaussian refinement are now preserved when two refined lines are concatenated.
```
```suggestion
* Parameters corresponding to Gaussian refinement are now preserved when two refined tracks are concatenated.
```

rpauszek commented on lumicks/pylake/kymotracker/detail/localization_models.py:
```diff
@@ -30,6 +30,21 @@ def _flip(self, size):
+                f"{other.__class__.__name__}."
+            )
+
+        init_args = {
```
nitpick: `init_kwargs`?

JoepVanlier replied:
👍

rpauszek commented on lumicks/pylake/kymotracker/kymotrack.py:
```diff
@@ -273,6 +277,18 @@ def interpolate(self):
+        node : int
+            Node index to split at
+        """
+        return self._with_coordinates(
+            self.time_idx[:node], self._localization[:node]
+        ), self._with_coordinates(self.time_idx[node:], self._localization[node:])
```
nitpick: I don't really like this enforced formatting. Maybe clearer to just make two variables and then return both rather than all in one shot

rpauszek commented on lumicks/pylake/kymotracker/kymotrack.py:
```diff
@@ -764,12 +780,23 @@ def _flip(self):
+        """
+        first_half, second_half = track._split(split_node)
+        self._src[self._src.index(track)] = first_half
+        self._src.append(second_half)
```
I wonder would it make more sense to insert the "new" track right after the first half of the original one? Basically if you start with `[track1, track2, track3]` and split on `track2` end up with `track1, track2a, track2b, track3]`. I think there are arguments for both. This way you still preserve the spatial order of the tracks (but that's not really guaranteed on our side anyway once you start messing the things), whereas your proposal would preserve the indexing of the list.

I could be convinced either way, just some food for thought

JoepVanlier replied:
Personally, I see more room for error by performing that insertion here. One might not realize they're performing an insertion, while iterating on a list and selectively splitting and then get errors.

rpauszek replied:
good point, yeah let's keep it as is

rpauszek commented on lumicks/pylake/nb_widgets/detail/mouse.py:
```diff
@@ -52,12 +54,17 @@ def set_active(self, state):
-                "motion_notify_event", lambda event: self.handle_motion(event)
-            )
+
+            if self._drag_callback:
```
why do we need this extra `if` now?

rpauszek commented on lumicks/pylake/nb_widgets/kymotracker_widgets.py:
```diff
@@ -169,6 +170,29 @@ def get_nearest(x, y):
+                # Explicit copy to ensure current state pushed to undo stack on assignment.
+                tracks = copy(self.tracks)
+
+                # If we click _beyond_ the note, cut there! Without adding the check on seconds
```
```suggestion
                # If we click _beyond_ the node, cut there! Without adding the check on seconds
```

rpauszek reviewed (approved):
looks good. left some minor suggestions/corrections. Also some thoughts on where we want to put the "new" track, but I can go either way

mikhas commented:
Nice work! So it splits the track between the two data points? What if I split between the first & second data point (or last & penultimate)? Or more general: What if the split results in tracks that suddenly don't make the `min_len` cut?

JoepVanlier commented:
> So it splits the track between the two data points? 
Yes.

> What if I split between the first & second data point (or last & penultimate)?
You get a single point track, which is invisible in the kymotracker. This is a good point.

> Or more general: What if the split results in tracks that suddenly don't make the `min_len` cut?
I would argue that a pass would have to be added to filter tracks `< min_length` after this split operation on the group.

rpauszek commented:
> > So it splits the track between the two data points?
> 
> Yes.
> 
> > What if I split between the first & second data point (or last & penultimate)?
> 
> You get a single point track, which is invisible in the kymotracker. This is a good point. I would argue split should either throw or be a `nullopt` if one of the sides ends up empty or a single point.
> 
> > Or more general: What if the split results in tracks that suddenly don't make the `min_len` cut?
> 
> I would argue that a pass would have to be added to filter tracks `< min_length` after this split operation on the group.

`min_length` is only enforced when you call `filter_tracks` so as far as the widget goes, this specific case should be handled there (not in `KymoTrackGroup`). I would then say probably `raise` if a `0` length track results, but allow for a length of `1`, since you can get these from `track_greedy()` anyways, and then need to specifically call `filter_tracks` afterwards.

JoepVanlier commented:
> > > So it splits the track between the two data points?
> > 
> > 
> > Yes.
> > > What if I split between the first & second data point (or last & penultimate)?
> > 
> > 
> > You get a single point track, which is invisible in the kymotracker. This is a good point. I would argue split should either throw or be a `nullopt` if one of the sides ends up empty or a single point.
> > > Or more general: What if the split results in tracks that suddenly don't make the `min_len` cut?
> > 
> > 
> > I would argue that a pass would have to be added to filter tracks `< min_length` after this split operation on the group.
> 
> `min_length` is only enforced when you call `filter_tracks` so as far as the widget goes, this specific case should be handled there (not in `KymoTrackGroup`). I would then say probably `raise` if a `0` length track results, but allow for a length of `1`, since you can get these from `track_greedy()` anyways, and then need to specifically call `filter_tracks` afterwards.

I added a `min_length` parameter to the split function at the group level. When tracks are being dropped, it now issues a warning which gets propagated to the widget. That didn't quite sit well with me for the edge cases where an empty track is returned however.

Just to compare, I added two options to ponder. The first is to just return fewer tracks from `split` when one is empty. This then gets filtered, and issues a warning to the user. Personally, I would find this confusing.

The second is to raise when one side is truly empty and then catch that in `_split_tracks()` at the group level. I think the latter is better, since it allows us to just handle the end points as "no operation took place", which I think is less confusing to users, than getting a warning but seeing no change on screen.

What do you think @mikhas @rpauszek ?

mikhas commented:
The warning in the first option feels odd because -- as you said yourself -- you already decided to handle the case gracefully, yet still issue a warning. So the question then becomes whether the warning is necessary.

In the second option, you say silently ignoring the edge cases, even _if_ they could be handled somewhat gracefully, from the side of the library implementer, is never OK. Seeing how the edge cases are at times difficult to reason about in an UI, I probably prefer the explicit nature of this option, even if it is slightly more inconvenient to use, API-wise.

So, what I ask myself now is whether KymoTracksGroup operations have grown esoteric enough that the use of context managers is warranted?

JoepVanlier commented:
> The warning in the first option feels odd because -- as you said yourself -- you already decided to handle the case gracefully, yet still issue a warning. So the question then becomes whether the warning is necessary.

Fair point. Especially considering it is an explicit argument to filter small tracks, it isn't.

> In the second option, you say silently ignoring the edge cases, even _if_ they could be handled somewhat gracefully, from the side of the library implementer, is never OK.

It is a private function whose sole purpose is to do something in the widget. Saying that though, did convince me that the logic should probably live in the widget.

> Seeing how the edge cases are at times difficult to reason about in an UI, I probably prefer the explicit nature of this option, even if it is slightly more inconvenient to use, API-wise.

Right. Explicit exceptions it is. I'll just let it bubble up. You can refer to the code in the kymotracker to see how it should be used in that case.

> So, what I ask myself now is whether KymoTracksGroup operations have grown esoteric enough that the use of context managers is warranted?

I don't really see how a context manager would be beneficial in this situation.

rpauszek reviewed (approved):
🚢

JoepVanlier commented on changelog.md:
```diff
@@ -19,6 +20,8 @@
 * Provided better initial guess for dwell time models with three or more components. Prior to this change, the dwelltime models involving three components or more would get an initial guess with an average lifetime beyond the mean of the observations.
 * Silenced false positive warnings about the optimization violating the bound constraints. Prior to this change, dwell time analysis would produce warnings about the optimizer exceeding bounds. These excursions beyond the bound are very small and have no effect on the model.
+* Parameters corresponding to Gaussian refinement are now preserved when two refined tracks are concatenated.
+* Set minimum `KymoTrack` length to `2` points.
```
Yes!

JoepVanlier commented on lumicks/pylake/kymotracker/kymotrack.py:
```diff
@@ -764,12 +792,35 @@ def _flip(self):
         flipped_kymo = self._kymo.flip()
         return KymoTrackGroup([track._flip(flipped_kymo) for track in self])
 
+    def _split_track(self, track, split_node, min_length):
```
I think in this case, especially since it's internal, it's better to not give it a default to force the user to think about it.
